### PR TITLE
objdetect: Fix sampling QR code image

### LINKED ISSUE #21929: Swiss QR-bills (ISO 20022) not decoded
##### System information (version)
- OpenCV =>  4.5.5.64
- Operating System / Platform => Ubuntu Linux
- Compiler => gcc?

##### Detailed description

The Swiss QR-bills are detected but not decoded.

See also: https://scanbot.io/en/sdk/scanner-sdk/barcode-scanner-sdk/swiss-qr

##### Steps to reproduce

```python
import cv2

image = cv2.imread("test-qr-code/155567196-54dbec58-7fa4-473a-a551-b52ecd04e11e.png")
detector = cv2.QRCodeDetector()
print(detector.detect(image))

data, bbox, straight_qrcode = detector.detectAndDecode(image)
print(data)
print(bbox)
```
With image: 155567196-54dbec58-7fa4-473a-a551-b52ecd04e11e
=>
```
(True, array([[[ 42.,  42.],
        [433.,  42.],
        [433., 433.],
        [ 42., 433.]]], dtype=float32))

[[[ 42.  42.]
  [433.  42.]
  [433. 433.]
  [ 42. 433.]]]
```

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found any solution
   
 - [x] I updated to the latest OpenCV version and the issue is still there
   
 - [x] There is reproducer code and related data files: videos, images, onnx, etc

### PR DESCRIPTION
**Pull request by Kumataro**
Fix #21929 

This patch replaces the delta_rows skip loop with a simpler version_size loop. 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [X] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by AleksandrPanov** [comment]
@Kumataro, need to add test. You could use sample from issue https://github.com/opencv/opencv/issues/21929.

File: modules/objdetect/test/test_qrcode.cpp [hunk 1]
```diff
@@ -669,6 +669,33 @@ TEST(Objdetect_QRCode_detect, detect_regression_20882)
...
+
+    Mat src;
+    cv::resize( qrimg, src, Size(200,200), 1.0, 1.0, INTER_NEAREST );
+
+    QRCodeDetector qrcode;
+    std::vector<Point> corners;
+    Mat straight_barcode;
+
+    EXPECT_TRUE(qrcode.detect(src, corners));
+    EXPECT_TRUE(!corners.empty());
```
Comment (asmorkalov) [inline modules/objdetect/test/test_qrcode.cpp:690]: It makes sense to check qr code coordinates as soon as it's synthetic and coordinates are well known.

**Comment by Kumataro** [comment]
Thank you for your comment!

I have been set up the development environment for OpenCV3.4.
I hope to fix it based on this environment within a few days.

But, In the current patch, each sampling area of cells are shifted to left/top a little, and the right/bottom pixels of the postIntermediate image are slightly truncated.

Further consideration and patches are needed.
(And testing detected corner  position is needed. )

It will take some time. I'm sorry.

**Comment by Kumataro** [comment]
I rewrote the test code prior to rebase to 3.4. I found that my patch wasn't a sufficient fix.

https://github.com/opencv/opencv/pull/22025 seems to contain a similar  fix at https://github.com/opencv/opencv/pull/21949. 

I think  it is better to close this pull to avoid a fix conflict.

Thank you very much!!

**Comment by alalek** [comment]
Rebased to 3.4.
We plan to merge this PR and then polish #22025

**Comment by alalek** [comment]
Sorry, #22025 has been merged first. Looks like all changes including test is already there.

**Pull request closed without merging**